### PR TITLE
Cache PullRequest Divergence

### PR DESCRIPTION
**Pull request by 6543**
taken from #10905

### CODE REVIEW

**Comment by 6543** [comment]
depend on #10905 ...

File: services/pull/check.go [hunk 1]
```diff
@@ -32,7 +32,13 @@ func AddToTaskQueue(pr *models.PullRequest) {
...
 		err := prQueue.PushFunc(strconv.FormatInt(pr.ID, 10), func() error {
 			pr.Status = models.PullRequestStatusChecking
-			err := pr.UpdateColsIfNotMerged("status")
+			divergence, err := GetDiverging(pr)
+			if err != nil {
+				log.Error("AddToTaskQueue.GetDiverging[%d]: %v", pr.ID, err)
+			}
+			pr.CommitsAhead = divergence.Ahead
+			pr.CommitsBehind = divergence.Behind
+			err = pr.UpdateColsIfNotMerged("status")
```
Comment (lafriks) [inline services/pull/check.go]: Shouldn't this so update divergence columns?

Comment (6543) [inline services/pull/check.go]:
```suggestion
			err = pr.UpdateColsIfNotMerged("status", "commits_ahead", "commits_behind")
```

**Comment by 6543** [comment]
Tested Speed:
**without**: Page: 250ms Template: 105ms
**cached**: Page: 129ms Template: 93ms

Test setup: 
* created a new test repo with initialisazion
* use webedit to edit a few lines to a new branch
* create pull request
* test page load speed of pull request

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
Don't know why but tests fail

**Comment by 6543** [comment]
Code Itself is ready: https://drone.gitea.io/go-gitea/gitea/23251

But if we like to have the "Update" button on old pull-request instead we have to extend the migration, else it will work fine after 1 push to base/head-branch

So if migration doesnt work stable, just drop it?

**Comment by 6543** [comment]
I think our test data is broken ... I'll try to fix it - could take a while

**Comment by 6543** [comment]
@lafriks found reason https://github.com/go-gitea/gitea/pull/10914/commits/39e12baf9eb6456b868e229e8cb375b2527db747

File: models/migrations/v136.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+		}
+		var results = make([]*models.PullRequest, 0, batchSize)
+		err := sess.Where("has_merged = ?", false).OrderBy("id").Limit(batchSize, last).Find(&results)
+		if err != nil {
+			return err
+		}
+		if len(results) == 0 {
+			break
+		}
+		last += len(results)
```
Comment (lunny) [inline models/migrations/v136.go:40]: A small performance optimiziation is to use `results.LoadIssues()` but not load issue in the loop.

File: models/pull.go [hunk 3]
```diff
@@ -352,6 +354,26 @@ func (pr *PullRequest) GetCommitMessages() string {
 	return stringBuilder.String()
 }
 
+// GetCommitDivergence get Divergence of a pull request
+func (pr *PullRequest) GetCommitDivergence() *git.DivergeObject {
+	return &git.DivergeObject{Ahead: pr.CommitsAhead, Behind: pr.CommitsBehind}
+}
+
+// UpdateCommitDivergence update Divergence of a pull request
+func (pr *PullRequest) UpdateCommitDivergence(div *git.DivergeObject) error {
```
Comment (lunny) [inline models/pull.go]: `models` depends on `git` is not a very good idea. I think we can just use `commitsAhead` and `commitsBehind` as two parameters of the method.

File: models/migrations/v136.go (hunk 2 above)
Comment (6543) [reply models/migrations/v136.go:40]: LoadIssues() for a PullList does not exist :(

File: models/pull.go (hunk 3 above)
Comment (6543) [reply models/pull.go]: fair point - change it

Comment (6543) [reply models/pull.go]: * `func (pr *PullRequest) GetCommitMessages() string {}`
* `func (pr *PullRequest) GetLastCommitStatus() (status *CommitStatus, err error) {}`
* `func (pr *PullRequest) IsHeadEqualWithBranch(branchName string) (bool, error) {}`

has to move then too - I'll send a pull for it later ...

**Comment by 6543** [comment]
@lunny pleace look at code comments :)

File: models/pull.go (hunk 3 above)
Comment (6543) [reply models/pull.go]: @lunny -> https://github.com/go-gitea/gitea/pull/11025

**Comment by lafriks** [comment]
Please resolve conflicts

**Comment by 6543** [comment]
done

**Comment by 6543** [comment]
@lunny do you have some time to look at it :pleading_face:

File: routers/repo/pull.go [hunk 4]
```diff
@@ -420,19 +418,13 @@ func PrepareViewPullInfo(ctx *context.Context, issue *models.Issue) *git.Compare
...
-			ctx.ServerError("IsUserAllowedToUpdate", err)
-			return nil
-		}
-		ctx.Data["UpdateAllowed"] = allowUpdate
-
-		divergence, err := pull_service.GetDiverging(pull)
-		if err != nil {
-			ctx.ServerError("GetDiverging", err)
-			return nil
+		if pull.CommitsBehind > 0 {
```
Comment (lunny) [inline routers/repo/pull.go]: Why we change the logic since the title said we will do some cache thing?

Comment (6543) [reply routers/repo/pull.go]: @lunny this is a smal **refactor**
-> why should we check if Update is Allowed if nothing can be Updated?

**Review by lunny — APPROVED** [review]

**Comment by 6543** [comment]
ready to :rocket:

**Pull request merged**